[RFC] Move part of syntax/vim.vim to syntax/vim/generated.vim

ZyX-I opened the pull request:
First two commits aim at providing necessary information in a lua-friendly format which does not require implementing parser. Commands are already in a friendly format, these commits move autocommands and options there as well.

Autocommand names move has additional benefit of avoiding duplicating data (enum and actual autocommands list; originally this commit was created because I needed list of autocommands for my parser).

Options were moved because I think that the result is more readable then the initial value; and it is also easier to convert from lua dictionary then from C structure (I constantly hear rumors about reworking option facility). Functions are taken directly from eval.c.

File is not actually generated or used because I don’t know which would be the correct variant for using it: creating `runtime/CMakeLists.txt` file or using the top-level one.

ZyX-I committed:
fileio: Move event definitions to the generator script

ZyX-I committed:
options: Move option definitions to options.lua

ZyX-I committed:
scripts: Add script which is able to generate part of vim.vim file

ZyX-I commented:
Build failed due to unrelated reason:

```
W: Failed to fetch http://www.rabbitmq.com/debian/dists/testing/main/i18n/Translation-en Unable to connect to www.rabbitmq.com:http:

E: Some index files failed to download. They have been ignored, or old ones used instead.

The command "if [ $TRAVIS_OS_NAME = linux ]; then sudo apt-mark hold oracle-java7-installer oracle-java8-installer; sudo apt-get update; elif [ $TRAVIS_OS_NAME = osx ]; then brew update; fi" failed and exited with 100 during .

Your build has been stopped.
```

. Will restart it.

ZyX-I commented:
There are also differences between generated keywords for commands and currently used keywords: I have not looked at all of them, but e.g. `loadk` is represented as `loadkeymap` and `loadk` in two adjacent lines (both variants are present in the middle) which is rather strange, my file uses `loadk[emap]` instead. Also note that some abbreviations do not exist officially (not mentined in `:h`), though they work and will be highlighted with the generated file.

splinterofchaos commented:
This definitely looks like an improvement to me. :+1: 

I wonder if a `runtime/doc` file shouldn't be generated to list the undocumented (and documented) abbreviations, just linking them to their non-abbreviated `:hep` sections.

ZyX-I commented:
This is not yet an improvement. I need to know where to put runtime file generation: either to `/runtime/CMakeLists.txt` or to `/CMakeLists.txt`.

ZyX-I commented:
Wondering what this may mean:

```
$ git clone --depth=50 git://github.com/neovim/neovim.git neovim/neovim
Cloning into 'neovim/neovim'...
fatal: read error: Connection reset by peer
The command "eval git clone --depth=50 git://github.com/neovim/neovim.git neovim/neovim" failed. Retrying, 2 of 3.
Cloning into 'neovim/neovim'...
remote: Counting objects: 3365, done.
remote: Compressing objects: 100% (2600/2600), done.
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated
```

```
$ git clone --depth=50 git://github.com/neovim/neovim.git neovim/neovim
Cloning into 'neovim/neovim'...
remote: Counting objects: 3365, done.
remote: Compressing objects: 100% (2600/2600), done.
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated
```

(first is gcc failure, second is gcc-32 failure).

ZyX-I commented:
Looks like build failed again due to DDoS or something like this.

ZyX-I committed:
cmake: Generate tags and some vim syntax elements

ZyX-I commented:
Since I got no replies about where I should place the runtime file generation it was placed to `runtime/CMakeLists.txt`. Help tags file generation was also moved there, and it is now part of the `make all` and not part of the installation step.

ZyX-I commented:
I have rebased on the current master. Is anybody going to merge this? With constant changes to various option defaults rebasing is not very fun: got 236 lines in `git diff -U0 autovimsyn..autovimsyn-old src/nvim/options.lua`, and there is no conversion script (not very trivial to write).

ZyX-I committed:
cmake: Run nvim that generates helptags with proper WORKING_DIRECTORY

justinmk commented:
I missed this somehow, so thanks for the reminder. Considering we've had zero problems since @ZyX-I converted the ex_cmds list to a lua generator, I have no compunctions about merging this.

justinmk commented on runtime/syntax/vim.vim:
```diff
@@ -64,24 +33,20 @@ syn match   vimTermOption contained	"t_&8"
+<<<<<<< HEAD
 syn keyword vimOnlyHLGroup contained	VisualNOS
 syn keyword nvimHLGroup contained	EndOfBuffer TermCursor TermCursorNC
+=======
```
bad merge

justinmk commented:
@ZyX-I There is a mismerge in the syntax file, but after that feel free to merge this.

ZyX-I commented:
Fixed mismerge.

ZyX-I commented:
Also note that for some reason I forgot to remove vimFuncName definitions from vim.vim. This was fixed alongside with the mismerge.

fmoralesc commented:
I don't think this handles the `vimOnly*` groups, which would cause some reversion (imho) from #2902.

ZyX-I commented:
@fmoralesc `vimOnly` groups will now be restored. The only thing which

ZyX-I committed:
syntax: Restore vim-specific highlight groups

ZyX-I committed:
syntax: List NeoVim-specific autocmd events

Note: list was copied from master (7a6bf3f418c5ad94ac2ac71f21275a87d08e87b9). 
I did not add more events there.

ZyX-I committed:
syntax: Exclude deprecated options

Excluded options (compatible, edcompatible and ttyfast) are listed in 
vimOnlyOption group.

fmoralesc commented:
@ZyX-I Great. LGTM, then ;)

Pyrohh commented:
I did an installation after this was merged, and noticed the directory structure has changed, was that intentional?
```
/usr/local/share/nvim $ tree -L 2
.
|-- doc
|   |-- makehtml.awk
|   `-- maketags.awk
|-- macros
|   `-- less.sh
`-- runtime
    |-- CMakeFiles
    |-- CMakeLists.txt
    |-- autoload
    |-- bugreport.vim
    |-- cmake_install.cmake
    |-- colors
    |-- compiler
    |-- delmenu.vim
    |-- doc
    |-- filetype.vim
    |-- ftoff.vim
    |-- ftplugin
    |-- ftplugin.vim
    |-- ftplugof.vim
    |-- indent
    |-- indent.vim
    |-- indoff.vim
    |-- keymap
    |-- macmap.vim
    |-- macros
    |-- makemenu.vim
    |-- menu.vim
    |-- mswin.vim
    |-- optwin.vim
    |-- plugin
    |-- print
    |-- rgb.txt
    |-- scripts.vim
    |-- synmenu.vim
    |-- syntax
    `-- vimrc_example.vim

15 directories, 22 files
/usr/local/share/nvim $
```

I'm not sure if all of these are issues:
- `less.sh` should be located in `$VIMRUNTIME/macros` according to `help :less`
- the following are duplicated from `runtime/`: `less.sh` and `make(html|tags).awk`
- CMakeFiles, CMakeLists.txt, and cmake_install.cmake probably shouldn't be installed

ZyX-I commented:
@Pyrohh You are right, I have leaked some errors in the cmake script.
